Add deterministic time to solver results

begunefeoglu opened the pull request:
## Fixes # .

## Summary/Motivation:
To be able to read deterministic solution time (in ticks) from Pyomo SolverResults

## Changes proposed in this PR:
- Added  `deterministic_time` to `SolverInformation`
- Updated CPLEX log file parser to get `deterministic_time`
- Added maxDetTimeLimit as `TerminationCondition`
- Updated CPLEX log file parser to get `maxDetTimeLimit`
- Added tests for `maxTimeLimit` and `maxDetTimeLimit`

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

blnicho committed:
Merge pull request #1268 from qtothec/gdpopt-update-19.1.22

GDPopt 20.02.28 updates

emma58 committed:
Removing redundant checks for _DisjunctData and SimpleDisjunct

blnicho committed:
Merge pull request #1285 from michaelbynum/pynumero_sparse

Pynumero.sparse Updates

emma58 committed:
Referencing set rewrite PR so we can remove one hack later

jsiirola committed:
Merge pull request #1322 from jsiirola/pypy-testing

Update tests to use sys.executable to launch python subprocesses

jsiirola committed:
Updating the ftoa tests to trigger loss of precision warning

jsiirola committed:
ftoa tests: handle platforms where np.longdouble == float

jsiirola committed:
Standardizing github actions naming

jsiirola committed:
Renaming 'parallel' to 'mpi'

jsiirola committed:
Updating checkout to @v2 to resolve rebuild issues

DLWoodruff committed:
Repair damage from a bug in tree_structure that manifests itself only… (#1321)

* Repair damage from a bug in tree_structure that manifests itself only in csvwriter

blnicho committed:
Merge pull request #1323 from jsiirola/github-actions-naming

Standardizing github actions naming

blnicho committed:
Merge pull request #1320 from qtothec/ftoa-bugfix

Add missing import for ftoa

e-dorigatti committed:
explained how to use indexed vars/constraints with persistent solver

e-dorigatti committed:
fixed variable name and removed superfluous comments

e-dorigatti committed:
improved variable names

michaelbynum committed:
Update persistent_solvers.rst

michaelbynum committed:
Merge pull request #1 from Pyomo/master

merge master into pynumero_mumps

jsiirola committed:
Merge pull request #1324 from e-dorigatti/persistent-indexed-doc

Documentation: indexed components in persistent interfaces

jsiirola committed:
Simplify MPI test skip logic

jsiirola committed:
Overhaul GitHub actions MPI workflow

- Update workflow to follow unix workflow.
- first attampt at adding codecov reporting to actions

jsiirola committed:
(debug) Temporarily enable branch push testing

jsiirola committed:
(debug) Temporarily enable branch push testing

jsiirola committed:
(debug) Temporarily enable branch push testing

jsiirola committed:
(debugging) MPI test driver

jsiirola committed:
(debugging) persist COVERAGE_PROCESS_START

jsiirola committed:
codecov: disable gcov

jsiirola committed:
Final cleanup of the MPI workflow driver

jsiirola committed:
Resolving test import errors when scipy is not present

jsiirola committed:
Correct import error when numpy is missing but mpi4py is present

michaelbynum committed:
pynumero requires numpy>=1.13

michaelbynum committed:
Merge pull request #5 from michaelbynum/mpi-test-coverage

pynumero requires numpy>=1.13

michaelbynum committed:
adding some information on working with forks and branches

michaelbynum committed:
Merge pull request #1325 from jsiirola/mpi-test-coverage

Improve testing and coverage reporting for MPI tests

jsiirola committed:
Adding an attempt_import() utility

jsiirola committed:
First step to move pynumero to use attempt_import

jsiirola committed:
Importing pyomo should import common (so logging is eset up)

jsiirola committed:
Cleaning up unnecessary pyomo.common imports

jsiirola committed:
Propagate the use of pyomo.common.dependencies for all yaml imports

This also suppresses loader deprecation warnings on newer versions of pyyaml.

michaelbynum committed:
Merge pull request #1314 from michaelbynum/pynumero_mumps

Pynumero MUMPS interface

jsiirola committed:
Adding missing import

fracek committed:
Change mul order in taylor series to avoid numpy compatibility issues

jsiirola committed:
Correcting guard around hasattr for conditional import

michaelbynum committed:
Merge pull request #1329 from fracek/taylor-series-floating-point

Change mul order in taylor series to avoid numpy compatibility issues

carldlaird committed:
Merge pull request #1328 from jsiirola/attempt-import

Add attempt_import(), standardize yaml imports

carldlaird committed:
updated document for working with remotes

carldlaird committed:
fixing headings and links

carldlaird committed:
Merge remote-tracking branch 'main-pyomo/master' into units-convert-pr-1

carldlaird committed:
Merge remote-tracking branch 'my-fork/units-convert-pr-1' into units-convert-pr-1

carldlaird committed:
change to get rid of import error

carldlaird committed:
Moved link to git book to the first paragraph of that section

jsiirola committed:
Rename FileDownloader.get_url() to get_platform_url()

jsiirola committed:
Add FileDownloader.get_text_file() method

jsiirola committed:
Adding tests

jsiirola committed:
Python 3.x fixes

michaelbynum committed:
Merge pull request #1330 from jsiirola/get-text-file

Get text file

michaelbynum committed:
Merge pull request #1326 from michaelbynum/docs_git

Docs for working on forks

jsiirola committed:
Updating documentation

emma58 committed:
Merge pull request #1129 from emma58/bugfixes-for-iterative-algs

BigM transformation rewrite

jsiirola committed:
Update MPI driver to generate parse_table_datacmds.py with a single process

michaelbynum committed:
Merge pull request #1331 from jsiirola/mpi-fix

Update GitHub Actions MPI driver

jsiirola committed:
Fix calling construct() on a constructed ContinuousSet

Robbybp committed:
flattener bug fix and test to catch

Robbybp committed:
delete commented line

carldlaird committed:
tested convert - working on model check

carldlaird committed:
Added helper function to retrieve solver success (or not)

carldlaird committed:
Added checks for locally and globally optimal as well

carldlaird committed:
changing name to optimal_termination from solve_optimal

jsiirola committed:
Don't extract the entire stack to get the caller's frame

carldlaird committed:
added exception option so this would be a one-line check for most users

kaklise committed:
parmest cleanup

kaklise committed:
parmest docs update

kaklise committed:
updated css override to wrap tables

kaklise committed:
renamed example files

Commit:
:rotating_light: Register factories using top-level import

Commit:
:rotating_light: Hardcode string results

carldlaird committed:
adding drop_units and model checking

carldlaird committed:
added US dollars (usd) as a currency by default

carldlaird committed:
adding ability to add new units definitions

qtothec committed:
shorten "short" names to 60 characters to allow for '_lb' and '_ub' suffixes

carldlaird committed:
supporting external functions and named expressions

jsiirola committed:
Merge branch 'master' into set-merge

jsiirola committed:
Merge branch 'master' into parmest

carldlaird committed:
Renaming according to Qis suggestion - now there are two methods - one that checks and one that asserts

carldlaird committed:
Merge remote-tracking branch 'main-pyomo/master' into units-convert-pr-1

carldlaird committed:
making consistent with dependencies.py

carldlaird committed:
cleaning up keyword argument testing in external function components

Commit:
Add deterministic time to solver results

Commit:
:books: Formatting of test results

Commit:
:books: Test docs

Commit:
:bug: Fix branching priorities for Kernel

ruaridhw commented:
Looks great @begunefeoglu! Two small things from my side:

1. Would it be possible to include a test by providing an example CPLEX log file? See `TestCPLEXSHELLProcessLogfile`
2. I don't think it should be too difficult to include this functionality in the Direct solver. See `_postsolve()` where `results.solver.wallclock_time` is set and `_apply_solver()` for how to wrap the `solve()` command. [This answer](https://stackoverflow.com/a/46837447/7312238) should also help.

carldlaird committed:
Merge pull request #1335 from carldlaird/solve_status_rapper

Added helper functions to retrieve solver success (or not)

ruaridhw commented:
I think it's an open question as to whether we should be retrieving the deterministic time from the logs or by wrapping `solve()`.

The latter is more consistent with what Pyomo have already done and what the CPLEX API seem to direct one towards. The former is in my view probably more accurate though as it is a number captured and reported by the solver itself rather than captured outside of the solve by an API call. What do you think?

blnicho committed:
Minor documentation edits

blnicho committed:
Temporarily disabling table wraping in docs

blnicho committed:
Merge branch 'master' into parmest

blnicho committed:
Enable table wrapping in the docs

blnicho committed:
Temporarily disable parmest graphics tests

blnicho committed:
Fixing a few typos, fixes #1339

blnicho committed:
Fixing a typo

blnicho committed:
Trying to resolve Github Actions test failures

jsiirola committed:
Overhaul of attempt_import to support deferred import checks

jsiirola committed:
Adding attempt_import tests

jsiirola committed:
this_file(): do not generate the entire stack to get caller frame

Commit:
Modified MPI testing for Benders cut generator

Commit:
Merge remote-tracking branch 'origin' into par_benders_test

jsiirola committed:
Support logical operations on import indicators; add tests

jsiirola committed:
Fix indicator operators when the first argument is already a bool

jsiirola committed:
Overhaul Pyomo imports to use pyomo.common.dependencies

jsiirola committed:
Fix using DeferredImportIndicator in a boolean context

carldlaird committed:
updates from WIP reviews

Commit:
:hammer: Add `iter_component()`

- Add iterator for looping over the data objects of either a `base` or `kernel` component

Commit:
:hammer: Use `iter_component()`

michaelbynum committed:
updating unix workflow

michaelbynum committed:
updating mpi workflow

michaelbynum committed:
updating benders tests

michaelbynum committed:
Merge branch 'osx_tests' into par_benders_test

michaelbynum committed:
benders updates

jsiirola committed:
Merge pull request #1342 from michaelbynum/osx_tests

Update Github workflow: OSX environment setup

jsiirola committed:
Remove 'extras' for Python3.6 on Appveyor due to conda-forge dependency issues

Commit:
merge into par_benders_test
Merge remote-tracking branch 'origin' into par_benders_test

Commit:
removing extra import

kaklise committed:
disable graphics test

jsiirola committed:
Add implicit alternate names for dotted imports

jsiirola committed:
Add option to specift an importer function

jsiirola committed:
Additional use of deferred imports across Pyomo

jsiirola committed:
Bugfix: missing argument definition

blnicho committed:
Merge pull request #1343 from jsiirola/appveyor-36-extras

Remove 'extras' for Python3.6 on Appveyor

blnicho committed:
Only skip parmest graphics test on OSX

blnicho committed:
Merge branch 'master' into parmest

ruaridhw committed:
:bug: Fix Py2 import error

blnicho committed:
Update parmest tests to not save generated figures

jsiirola committed:
Re-adding pyomo.network.decomposition.imports_available

jsiirola committed:
Fix typo

Commit:
Add maxDetTimeLimit as TerminationCondition and test

michaelbynum committed:
Merge branch 'master' into par_benders_test

michaelbynum committed:
Merge remote-tracking branch 'cpmuir/par_benders_test' into par_benders_test

michaelbynum committed:
updating mpi workflow

michaelbynum committed:
adding checks for solvers for the benders tests

michaelbynum committed:
updating benders tests

cpmuir committed:
removing extra files

cpmuir committed:
removing extra files

cpmuir committed:
removing extra files

jsiirola committed:
Merge pull request #1300 from flexciton/cplex_branching_priorities

Add branching priorities to CPLEXSHELL

jsiirola committed:
Merge pull request #1338 from qtothec/gams_shorter_names

Shorten GAMS ShortNameLabeler symbols

jsiirola committed:
Fix pyutilib reference

jsiirola committed:
Fix DataPortal table generator (include param header)

This fixes the DataPortal table generator so that a header row is always
generated (either using the column names provided by the user or an
auto-generated set of column names).

michaelbynum committed:
simplifying benders tests

jsiirola committed:
Merge branch 'master' into param-domain

carldlaird committed:
clean up the API

Robbybp committed:
change sliced indices to list

jsiirola committed:
Have _ImplicitAny keep a reference to the owning Param

This allows the deprecation warning to inform the user which Param is
generating the warning.

jsiirola committed:
Provide expliciy domain for Params in Scenario Tree Model

jsiirola committed:
Avoid weakref in _ImplicitAny

jsiirola committed:
Resolve _ImplicitAny weakref pickle bug

cpmuir committed:
Merge pull request #3 from michaelbynum/par_benders_test

Simplifying benders testing

kaklise committed:
added assert statements to check data types

kaklise committed:
bug fix in confidence_region_test and added assert statements to check data type

kaklise committed:
Added additional tests

jsiirola committed:
Update codecov config to allow overall coverage to drop

This allows overall coverage to drop (due to code cleanups, or CI
provider lag / unavailability) without failing a PR.  PR patch coverage
must still be at the level of the overall codebase.

jsiirola committed:
bugfix for _ImplicitAny.__getstate__

jsiirola committed:
Adding guards for "None" weakrefs

jsiirola committed:
Adding Param domain deprecation test

jsiirola committed:
Fixing typos

jsiirola committed:
Merge pull request #1334 from Robbybp/flattener_bug

DAE model flattener bug fix

blnicho committed:
Merge pull request #1266 from jsiirola/param-domain

Deprecate 'Any' being the default Param domain.

blnicho committed:
Merge pull request #1345 from jsiirola/codecov-config

Update codecov config to allow overall coverage to drop

blnicho committed:
Merge pull request #1337 from kaklise/parmest

Parmest updates, including options for leave-N-out and data reconciliation

jsiirola committed:
Merge branch 'master' into import-speedup

begunefeoglu commented:
I am happy with the current solution, I don't see any benefit of more involved solution.

blnicho committed:
Merge pull request #1344 from jsiirola/import-speedup

Speed up import of pyomo.environ

jsiirola committed:
Merge pull request #1278 from cpmuir/par_benders_test

Benders MPI logic bug and parallel test coverage

jsiirola committed:
Merge branch 'master' into units-convert-pr-1

jsiirola committed:
Resolve units_container warnings and doctest errors

jsiirola committed:
Fixing tests (attempt_import defers by default)

blnicho committed:
Updating the CHANGELOG for the 5.6.9 release

jsiirola committed:
Finalizing the 5.6.9 release

blnicho committed:
Merge pull request #1350 from Pyomo/finalize-release

Finalize 5.6.9 release

jsiirola committed:
Excluding matplotlib 3.2.1 from allowable versions

jsiirola committed:
Preparing master for development (5.6.10.dev0)

blnicho committed:
Merge pull request #1352 from Pyomo/finalize-release

Preparing master for development (5.6.10.dev0)

Commit:
Add deterministic time to direct solver

jsiirola committed:
Pinning numpy, pandas versions

jsiirola committed:
Freeing matplotlib, numpy versions

jsiirola committed:
Relaxing pandas version restriction

jsiirola committed:
Documenting pandas version restriction

ruaridhw commented on pyomo/solvers/plugins/solvers/cplex_direct.py:
```diff
@@ -191,6 +192,7 @@ def _process_stream(arg):
 
             t1 = time.time()
             self._wallclock_time = t1 - t0
+            self._deterministic_time = self._solver_model.get_dettime()
```
@begunefeoglu I believe `get_dettime()` just returns a [timestamp](https://www.ibm.com/support/knowledgecenter/SSSA5P_12.7.0/ilog.odms.cplex.help/refjavacplex/html/ilog/cplex/IloCplex.html#getDetTime()), not the actual result in ticks.

> To measure deterministic time spent between a starting point and ending point of an operation, take the time stamp at the starting point; take the time stamp at the ending point; subtract the starting time stamp from the ending time stamp to get elapsed time in deterministic ticks.
>
> The absolute value of the deterministic time stamp is not meaningful.



Note this can also be copied to the PersistentSolver

begunefeoglu replied:
Good catch, thanks!

Commit:
correct how to measure det time

Commit:
Add deterministic time to persistencesolver

Commit:
typo

ruaridhw commented on pyomo/solvers/plugins/solvers/persistent_solver.py:
```diff
@@ -377,12 +377,16 @@ def solve(self, *args, **kwds):
 
             # we're good to go.
             initial_time = time.time()
+            initial_det_time = self._solver_model.get_dettime()
```
Unfortunately I don't think this will work...

`get_dettime()` is only exposed by the CPLEX solver API however this class is intended to be general for all possible PersistentSolvers (eg. Gurobi) as well which won't have this same interface.

ruaridhw commented on pyomo/solvers/tests/checks/test_cplex.py:
```diff
@@ -333,5 +333,41 @@ def test_log_file_shows_presolve_infeasible(self):
+        self.assertEqual(results.solver.status, SolverStatus.ok)
+        self.assertEqual(
+            results.solver.termination_condition, TerminationCondition.maxDetTimeLimit
+        )
```
Can we assert on the value(s) of `results.solver.deterministic_time` in these tests as well?

ruaridhw commented on pyomo/solvers/tests/checks/test_CPLEXDirect.py:
```diff
@@ -255,13 +255,13 @@ def test_soln_limit_mip(self):
                              TerminationCondition.unknown)
             self.assertEqual(model.solutions[0].status,
                              SolutionStatus.stoppedByLimit)
+            self.assertGreater(results.solver.deterministic_time, 0.0)
```
Perhaps a slightly stronger test would be to assert on the value itself?

I think it would be enough to just duplicate this test and pass in `opt.options['dettimelimit'] = 10` instead of `mip_limits_solutions`. Depending on how CPLEX behaves, we may need to assert that `results.solver.deterministic_time` is within a few ticks of 10; I'm not entirely sure whether it does stop exactly at the limit or whether it will perform some small cleanup as well.

begunefeoglu replied:
As you said, `results.solver.deterministic_time` returns `0.0009851455688476562` when I set `opt.options['dettimelimit'] = 0.0001`

ruaridhw replied:
Hmmm that's interesting... Maybe it would help to increase this to 10? 1 tick is actually a very short amount of time because I can see from the logs that CPLEX typically runs at several 100's of ticks _per second_.

ruaridhw replied:
Also this test is failing for a couple of the pipelines so that proves that the `deterministic_time` is non-deterministic _across_ platforms. As described above, I think using a much larger magnitude of 10 and asserting within a tolerance of 9.5-10.5 ticks should alleviate this.

begunefeoglu replied:
Optimal solution can be found in 0.0015420913696289062 ticks for this super tiny problem, that's why I set 0.0001 (hoping to hit the time limit and get `results.solver.deterministic_time=0.001`. The best solution I can think of is to assert within a tolerance of 0-0.001 ticks under the setting of `opt.options['dettimelimit'] = 0.001`

begunefeoglu replied:
Also, the test failed in the pipelines with `0.0029306411743164062 != 0.0009851455688476562`. (`0.0009851455688476562` was the result I got locally.) I think deterministic time might also be CPLEX version dependent which is something we cannot control. I prefer not to assert any UB.

ruaridhw reviewed (changes requested):
@begunefeoglu looks great, just a couple of minor things left.

Commit:
tests updated

jsiirola committed:
Make GAMSDirect.available() more robust to unexpected exceptions

jsiirola committed:
Remove pandas version restriction

The fix to GAMSDirect.available() should resolve test errors.

jsiirola committed:
Merge pull request #1351 from jsiirola/py38-gams-build-issue

Update GAMSDirect.available() to catch unexpected exceptions

jsiirola committed:
Merge branch 'master' into units-convert-pr-1

jsiirola committed:
Merge branch 'master' into set-merge

jsiirola committed:
Fixing _ImplicitAny for new set system

jsiirola committed:
Fix get_text_file() in Python 3.x

jsiirola committed:
Fix mock retrieve_url() to return correct types

jsiirola committed:
Fix setup.py: SystemExit.message doesn't exist in Python 3

Fixes #808, fixes #1118

michaelbynum committed:
Merge pull request #1354 from jsiirola/setup-systemexit-fix

Fix setup.py: SystemExit.message doesn't exist in Python 3

michaelbynum committed:
Merge pull request #1353 from jsiirola/get-text-file

Resolve Py3 issue in FileDownloader.get_text_file()

jsiirola committed:
Support multi-line comments in DAT lexer

jsiirola committed:
Fix DAT lexer to recognize/convert numbers

jsiirola committed:
Improve string tokenization

- WORD and WORDWITHLBRACKET must begin with an alpha character
- general STRING values will be automatically quoted
- QUOTEDSTRING and STRING values will NOT have the quote character removed

jsiirola committed:
Resolve escaped quotation (doubled ['"]) characters

jsiirola committed:
Simplify/speed up token processing

jsiirola committed:
Add tests for param data processing

jsiirola committed:
Fix numeric token processing for non-DAT sources

carldlaird committed:
initial version of interface between external input output model and pyomo with CyIpopt and PyNumero

carldlaird committed:
cleaning up documentation

carldlaird committed:
changed evaluate_derivatives to return a sparse matrix

carldlaird committed:
changing the documentation

jsiirola committed:
Clarifying re patterns for comments and quoted strings

Commit:
return solver-dependent deterministic time

Commit:
Add test for cplex direct solver

Commit:
reverted the changes in persistentsolver

carldlaird committed:
responding to PR comments

ruaridhw commented on pyomo/solvers/tests/checks/test_CPLEXDirect.py:
```diff
@@ -262,7 +263,27 @@ def test_soln_limit_mip(self):
                              SolutionStatus.stoppedByLimit)
-            self.assertGreater(results.solver.deterministic_time, 0.0)
+
+    def test_dettime_limit_mip(self):
```
@begunefeoglu I think we need to wrap this in the same `@unittest.skipIf` decorator otherwise the pipeline will fail if CPLEX is not available

Commit:
added test decorator

carldlaird committed:
changing to use unique_component_name

carldlaird committed:
adding copyright

jsiirola committed:
Merge pull request #1341 from carldlaird/units-convert-pr-1

Units support for Pyomo components and models

carldlaird committed:
Merge pull request #1356 from carldlaird/external-cyipopt-example

External CyIpopt Interface

jsiirola committed:
Removing dependence on pyutilib.ply

mrmundt committed:
Adding coverage to unix tests

mrmundt committed:
Fixing empty spaces

mrmundt committed:
Needed to add coverage install

mrmundt committed:
Removing local test action and cleaning up

jsiirola committed:
Addressing typos and review comments

jsiirola committed:
Propagating optimization from is_continuous to is_binary, is_integer

mrmundt committed:
Started failing - adding GAMS DIRs which may have caused the issue.

mrmundt committed:
Removing local testing

blnicho committed:
Updating the ContinuousSet docstring to match new behavior of enforcing bounds

jsiirola committed:
Minor formatting changes to consolidate unix & mpi drivers

mrmundt committed:
Removing any references/fixes for <Py2.7

mrmundt committed:
Merge pull request #1 from jsiirola/unix_coverage

Minor formatting changes to consolidate unix & mpi drivers

blnicho committed:
Merge pull request #1361 from mrmundt/remove_under_py27

Remove remaining < Python 2.7 references

ruaridhw commented:
The pipeline appears to be failing for all Windows jobs. I've raised #5 to see if it fails on a fresh branch too. Otherwise happy to merge.

jsiirola committed:
Moving linux to the ubuntu-latest image

jsiirola committed:
Fixing linux job name

jsiirola committed:
Renaming job names to match new Travis/Appveyor names

jsiirola committed:
Unix/OSX PR builds should also build the master branch

jsiirola committed:
Removing unnecessary code

jsiirola committed:
Removing unneeded configuration

jsiirola committed:
Adding AML GlobalSets to pyomo.kernel

This restores global set (Reals, etc) imports for kernel users through
the pyomo.kernel environment.  Reverting changes to the kernel variables
documentation and example.

jsiirola committed:
Updating Travis badge (migration from org to com)

jsiirola committed:
Add "continue on error" for codecov report

jsiirola committed:
Updating job names, pass job name to codecov

carldlaird committed:
Merge pull request #1364 from jsiirola/travis-migration

Updating Travis badge (migration from org to com)

jsiirola committed:
Merge pull request #1359 from mrmundt/unix_coverage

Add test coverage to linux/osx GitHub actions

jsiirola committed:
Removig trailing whitespace

blnicho committed:
Merge pull request #1366 from jsiirola/dat-fixes

Resolve DAT tokenizer bugs

blnicho committed:
Merge pull request #1319 from jsiirola/set-merge

Integrate new Set rewrite into Pyomo

Commit:
Merge remote-tracking branch 'pyomo_github/master'

ruaridhw commented:
This is great, thanks @begunefeoglu! The pipeline is passing with `Pyomo/pyomo/master` merged  in however as discussed offline, this branch actually contains quite a few API changes. Given that we're satisfied this PR passes the pipeline, for now, I'm going to merge #5 instead, close this PR, and ignore the fact that our `master` pipeline will now be breaking until we merge `Pyomo/pyomo/master` properly 😉
